tests/jdom/tester.ts

```diff
@@ -0,0 +1,191 @@
+        const busDevice = bus.addServiceProvider(new JDServiceProvider([device.server]))
+        return {
+            server: device.server,
+            busDevice: busDevice,
+            roleName: device.roleName,
+        }
+    })
+
+    bus.start()
+
+    // Wait for created devices to be announced, so services become available
+    const serverDeviceIds = serverDevices.map(elt => {
+        return elt.busDevice.deviceId
+    })
+    await new Promise(resolve => {
+        const devicesIdSet = new Set(serverDeviceIds)
+        const announcedIdSet = new Set()
+        const onHandler = (device: JDDevice) => {
+            if (devicesIdSet.has(device.deviceId)) {
+                announcedIdSet.add(device.deviceId)
+            }
+            if (setEquals(devicesIdSet, announcedIdSet)) {
+                bus.off(DEVICE_ANNOUNCE, onHandler)
+                resolve(undefined)
+            }
+        }
+        bus.on(DEVICE_ANNOUNCE, onHandler)
+    })
+
+    // Bind services to roles
+    // Start by geting server service -> roleName mappings
+    const serverServiceRoles = serverDevices.map(elt => {
+        const services = elt.busDevice.services({serviceClass: elt.server.serviceClass})
+        assert(services.length > 0, 
+            `created device ${elt.busDevice.friendlyName} has no service of ${elt.server.specification.name}`)
+        assert(services.length == 1, 
+            `created device ${elt.busDevice.friendlyName} has multiple service of ${elt.server.specification.name}`)
+        return {
+            service: services[0],
+            roleName: elt.roleName
+        }
+    })
+
+    const roleBindings = serverServiceRoles.filter(elt => {
+        return !!elt.roleName  // filter for where role name is available
+    }).map(elt => {
+        return {
+            role: elt.roleName, 
+            serviceClass: elt.service.specification.classIdentifier,
+            service: elt.service,
+        }
+    })
+
+    const roleManager = new RoleManager(bus)
+    roleManager.setRoles(roleBindings)
+
+    // Return created services as a map from the source server
+    // Trace devices are ignored here, since handles aren't (currently?) useful
+    const serviceMap = new Map(
+        serverDevices.map(elt => {
+            return [elt.server, elt.busDevice.services({serviceClass: elt.server.serviceClass})[0]]
+        })
+    )
+
+    // Actually run the test here
+    await test(bus, serviceMap)
+
+    bus.stop()
+}
+
+
+interface EventWithinOptions {
+    after?: number  // event must happen at least this many ms after the current time (by default, 0)
+    within?: number  // event must happen within this many ms after the current time (by default, infinite)
+    tolerance?: number  // when after is set, sets the allowable range for the event to be tolerance on either side of after
+                        // is an error if within is set, or after is not set
+}
+
+
+export class JDBusTestUtil {
```

Comment:
we don't really need a class here. you can get a grab of the bus from service.device.bus. nextEventFrom, nextEventWithin then become exported functions.

Reply:
Is the proposal to have these be top-level functions instead of being nested in a class?

Reply:
Correct.

Reply:
Done.